packages/thyra/core/src/support/operator_solve/client_support/Thyra_DefaultBlockedTriangularLinearOpWithSolve_def.hpp

```diff
@@ -550,11 +550,11 @@ DefaultBlockedTriangularLinearOpWithSolve<Scalar>::solveImpl(
     *this, M_trans, *X_inout, &B_in
     );
   TEUCHOS_TEST_FOR_EXCEPT(!this->solveSupportsImpl(M_trans));
-  TEUCHOS_TEST_FOR_EXCEPTION(
-    nonnull(solveCriteria) && !solveCriteria->solveMeasureType.useDefault(),
-    std::logic_error,
-    "Error, we can't handle any non-default solve criteria yet!"
-    );
+  // TEUCHOS_TEST_FOR_EXCEPTION(
```

Comment:
This is technically not correct.  This class can't just assume that every solve type is supported.    Instead, it will need to query each LinearOpWithSolveBase object on diagonal and ask them if they support that solve type.   Unit tests will take some work to test the various use cases.

Reply:
Why does this class even need to check this?  Shouldn't each diagonal solver check whether it is given a valid solve criteria, and throw if it wasn't.  I don't see why this class needs to duplicate those checks.

Reply:
> Why does this class even need to check this? Shouldn't each diagonal solver check whether it is given a valid solve criteria, and throw if it wasn't. 

Because this is part of the preconditions for calling `Thyra::DefaultBlockedTriangularLinearOpWtihSolve::solveImple()`.  

I think we need to think about what SolveCriteria means in the context of a composite solver object like this.  If you look at the spec for the [SolveCriteria](https://trilinos.org/docs/dev/packages/thyra/doc/html/structThyra_1_1SolveCriteria.html#details) struct, a genetic solver object like Thyra::DefaultBlockedTriangularLinearOpWithSolve cannot easily compute the types of global norms that are specified there except for perhaps simple ones like ||r||/||r0|| and it can do litter to dive the individual block solves to achieve this in any type of rigorous way.  I think a genetic class like this really can't say much about global norms like this without a lot of expense.  Therefore, the correct thing for it to do is to return a [SolveStatus](https://trilinos.org/docs/dev/packages/thyra/doc/html/structThyra_1_1SolveStatus.html#details) with `SOLVE_STATUS_UNKNOWN` for the `solveStatus` enum an `unkownTolerance()` for the `achievedTol` member.  That suggests that the `accumulateSolveStatus()` function should likely be updated to reflect this.

Why is Tempus even specifying a non-default SolveCriteria object?  And, how does a triangular operator strictly honor a non-default solve criteria?

Reply:
Tempus isn't specifying a SolveCriteria object, NOX is.  

This particular use in Tempus is actually fairly trivial where it is a 2x2 block diagonal system.  The first diagonal block is an arbitrary LOWS but the second is actually just a scaled identity operator with a trivial solve.  I was just trying to reuse some functionality from Thyra so that I didn't have to write my own LOWS for this case (although I ended up having to write a LOWSF anyway because the one in Thyra doesn't work in this case).  

I am not interested in making this class work in all situations.  I had to make these modifications to get it to not throw when called by NOX and for NOX to not print copious warnings about failed linear solves because the solve status wasn't being initialized.

As you point out, there are lots of combinations of solvers where this is problematic, but I'm not interested in making this code work in those cases because it didn't work in them before anyway. 

How about the following:
```
#ifdef THYRA_DEBUG
  THYRA_ASSERT_LINEAR_OP_MULTIVEC_APPLY_SPACES(
    "DefaultBlockedTriangularLinearOpWithSolve<Scalar>::apply(...)",
    *this, M_trans, *X_inout, &B_in
    );
  TEUCHOS_TEST_FOR_EXCEPT(!this->solveSupportsImpl(M_trans));
  TEUCHOS_TEST_FOR_EXCEPT(!this->solveSupportsMeasureTypeImpl(M_trans, 
      solveCriteria->solveMeasureType()));
#endif // THYRA_DEBUG  

  //
  // Y = alpha * inv(op(M)) * X + beta*Y
  //
  // =>
  //
  // X[i] = inv(op(Op[i]))*B[i], for i=0...numDiagBlocks-1
  //
  // ToDo: Update to handle off diagonal blocks when needed!
  //

  const ProductMultiVectorBase<Scalar>
    &B = dyn_cast<const ProductMultiVectorBase<Scalar> >(B_in);
  ProductMultiVectorBase<Scalar>
    &X = dyn_cast<ProductMultiVectorBase<Scalar> >(*X_inout);

  SolveStatus<Scalar> solveStatus;
  accumulateSolveStatusInit(outArg(solveStatus));
  for ( int i = 0; i < numDiagBlocks_; ++ i ) {
    const RCP<const LinearOpWithSolveBase<Scalar> >
      Op_k = diagonalBlocks_[i].getConstObj();
    Op_k->setOStream(this->getOStream());
    Op_k->setVerbLevel(this->getVerbLevel());
     SolveStatus<Scalar> status =
       Thyra::solve( *Op_k, M_trans, *B.getMultiVectorBlock(i),
                     X.getNonconstMultiVectorBlock(i).ptr(), solveCriteria );
     accumulateSolveStatus(*solveCriteria, status, solveStatus);
  }

  return solveStatus;
```

Reply:
@etphipp,

Actually, now that I think about this some more, I think your current implementation is more technically correct than the the one you show above using the current `accumulateSolveStatus()` implementation.  What the current implementation does is it returns `SOLVE_STATUS_CONVERGED` if all of the block solvers return `SOLVE_STATUS_CONVERGED` but it leaves the default value of `solveStatus.achievedTol` at the default value which is `SolveStatus::unknownTolerance()`.  In this case, that is the only generic tolerance you can return.  So if NOX is happy with that SolveStatus, then let's go with that.  We can worry about other the technicalities later if someone complains.  Therefore, go ahead and push what you have.  I think the current unit tests for that class will check this new implementation to some extent. 

FYI: I came to this realization when I thought about this use case some more and thought more about the implementation of the function `accumulateSolveStatus()`.  That implementation is really only correct for a very narrow set of cases.  Then I happened to actually read the user documentation for the function `accumulateSolveStatus()` shown at:

* https://trilinos.org/docs/dev/packages/thyra/doc/html/structThyra_1_1SolveStatus.html#aca3dea7982de16447ca8c521ef659d3f

That documentation says that this implementation is really only designed to accumulate the solve status objects for multiple RHS of the same linear operator.  For that use case, I think the implementation of that function is exactly mathematically correct, as defined by the [LinearOpWithSolveBase interface spec](https://trilinos.org/docs/dev/packages/thyra/doc/html/classThyra_1_1LinearOpWithSolveBase.html#LOWSB_solve_status_sec).  In fact, the spec is defined in by what is in that function!  But it is not really correct to accumulate the SolveStatuses of the blocks of a block triangular operator.  Therefore, we (likely I) should have called this function `accumulateMultiRhsSolveStatus()` instead of `acumulateSolveStatus()` and I would not have gotten confused.  (An example of where a longer name can be better to avoid confusion.)

Sorry to hold you up.

Thanks!